www/src/components/repository/RepositoryEdit.tsx

```diff
@@ -0,0 +1,677 @@
+      width={64 + 32}
+      height={64 + 32}
+      objectFit="cover"
+      backgroundColor="fill-one"
+      cursor="pointer"
+      border="1px solid border"
+      onClick={onClick}
+    />
+  )
+}
+
+type FormState = {
+  name: string;
+  description: string;
+  category: string;
+  oauthUrl: string;
+  oauthMethod: string;
+  tags: { name: string }[];
+  private: boolean;
+};
+
+function RepositoryEdit() {
+  const {
+    id,
+    name,
+    description,
+    category,
+    editable,
+    oauthSettings,
+    tags,
+    icon,
+    private: privateRepo,
+  } = useContext(RepositoryContext) as any
+  const {
+    state: formState,
+    initialState: formInitialState,
+    hasUpdates: formStateHasUpdates,
+    update: updateFormState,
+    // reset: resetFormState,
+  } = useUpdateState<FormState>(
+    useMemo(
+      () => ({
+        name,
+        description,
+        category,
+        oauthUrl:
+          typeof oauthSettings?.uriFormat === 'string'
+            ? oauthSettings?.uriFormat
+            : '',
+        oauthMethod:
+          typeof oauthSettings?.authMethod === 'string'
+            ? oauthSettings?.authMethod
+            : authMethods.BASIC,
+        tags: isArray(tags) ? tags.map(tag => ({ name: tag.name })) : [],
+        private: !!privateRepo,
+      }),
+      [
+        name,
+        description,
+        oauthSettings?.authMethod,
+        oauthSettings?.uriFormat,
+        tags,
+        privateRepo,
+        category,
+      ]
+    )
+  )
+
+  const [tagSearchString, setTagSearchString] = useState('')
+  const newTag = tagSearchString
+    .replaceAll(/([\s_]+)/gu, '-')
+    .replaceAll(/[^A-Za-z-]+/gu, '')
+    .toLowerCase()
+  const tagSearch = useQuery(TAGS_SEARCH_QUERY, {
+    variables: { q: tagSearchString, first: 200 },
+  })
+  const tagSearchResults: { tag: string; count: number }[] =
+    tagSearch?.data?.tags?.edges?.map((edge: any) => edge?.node) || []
+
+  const [iconUpdate, setIconUpdate] = useState<{
+    file: File | null;
+    previewUrl: string | null;
+  }>({ file: null, previewUrl: icon || null })
+
+  const [, setSuccess] = useState(false)
+  const [, setError] = useState<ApolloError>()
+
+  const tagSearchRef = useRef<any>(null)
+
+  const [mutation, { loading }] = useMutation(UPDATE_REPOSITORY_MUTATION, {
+    variables: {
+      repositoryId: id,
+      attributes: {
+        name: formState.name,
+        description: formState.description,
+        category: formState.category,
+        oauthSettings:
+          formState.oauthUrl && formState.oauthMethod
+            ? {
+              uriFormat: formState.oauthUrl,
+              authMethod: formState.oauthMethod,
+            }
+            : null,
+        ...(iconUpdate.file ? { icon: iconUpdate.file } : {}),
+        tags: formState.tags.map(t => ({ tag: t.name })),
+        private: formState.private,
+      },
+    },
+    update: (cache, { data: { updateRepository } }) => {
+      const prev: any = cache.readQuery({
```

Comment:
is this cache update necessary, my guess is apollo will just handle it?  If it is, we have a helper `updateCache` that could compress some of the logic as well